### PR TITLE
Set node name from loc in instrumentation tool.

### PR DESCRIPTION
**Pull request by imaihal**
As described in PR #2115, node name attribute is often lost especially when the operation is rewritten. So, node name was not often shown in output of instrumentation tool.

This PR fixes it by setting node name from location implemented in PR #2115.

- Example in roberta-base-11 model
  - onnx_node_name attribute in SoftmaxV11 is lost by rewriting to Softmax. So, node name was lost in output, but it appears after this PR.

Input (.onnx or .onnx.mlir)
- SoftmaxV11 has onnx_node_name.
```
 %373 = "onnx.Div"(%371, %372) {onnx_node_name = "Div_90"} : (tensor<*xf32>, tensor<f32>) -> tensor<*xf32>
 %374 = "onnx.Add"(%373, %303) {onnx_node_name = "Add_91"} : (tensor<*xf32>, tensor<*xf32>) -> tensor<*xf32>
 %375 = "onnx.SoftmaxV11"(%374) {axis = 3 : si64, onnx_node_name = "Softmax_92"} : (tensor<*xf32>) -> tensor<*xf32>
 %376 = "onnx.MatMul"(%375, %369) {onnx_node_name = "MatMul_93"} : (tensor<*xf32>, tensor<*xf32>) -> tensor<*xf32>
```

Output(Main branch):
- The onnx_node_name is lost when SoftmaxV11 is rewritten to Softmax. So, node name is not shown.
```
#678) before onnx.Div Time elapsed: 0.000001 accumulated: 1680759914.813203 (Div_90)
#679) after  onnx.Div Time elapsed: 0.000670 accumulated: 1680759914.813873 (Div_90)
#680) before onnx.Softmax Time elapsed: 0.000001 accumulated: 1680759914.813874
#681) after  onnx.Softmax Time elapsed: 0.006557 accumulated: 1680759914.820431
#682) before onnx.MatMul Time elapsed: 0.000001 accumulated: 1680759914.820432 (MatMul_93)
#683) after  onnx.MatMul Time elapsed: 0.004311 accumulated: 1680759914.824743 (MatMul_93)
```
Output(PR) (.onnx)
- Node name is set from NameLoc.
```
#678) before onnx.Div Time elapsed: 0.000001 accumulated: 1681981634.161860 (Div_90)
#679) after  onnx.Div Time elapsed: 0.001742 accumulated: 1681981634.163602 (Div_90)
#680) before onnx.Softmax Time elapsed: 0.000000 accumulated: 1681981634.163602 (Softmax_92)
#681) after  onnx.Softmax Time elapsed: 0.008226 accumulated: 1681981634.171828 (Softmax_92)
#682) before onnx.MatMul Time elapsed: 0.000001 accumulated: 1681981634.171829 (MatMul_93)
#683) after  onnx.MatMul Time elapsed: 0.005654 accumulated: 1681981634.177483 (MatMul_93)
```
Output (PR) (.onnx.mlir)
- Filename and line are used instead since NameLoc is not available in .onnx.mlir file. 
```
#534) before onnx.Div Time elapsed: 0.000001 accumulated: 1682993862.980648 (Div_90)
#535) after  onnx.Div Time elapsed: 0.000584 accumulated: 1682993862.981232 (Div_90)
#536) before onnx.Add Time elapsed: 0.000000 accumulated: 1682993862.981232 (Add_91)
#537) after  onnx.Add Time elapsed: 0.001126 accumulated: 1682993862.982358 (Add_91)
#538) before onnx.Softmax Time elapsed: 0.000000 accumulated: 1682993862.982358 (roberta-base-11.onnx.mlir:378)
#539) after  onnx.Softmax Time elapsed: 0.006471 accumulated: 1682993862.988829 (roberta-base-11.onnx.mlir:378)
#540) before onnx.MatMul Time elapsed: 0.000001 accumulated: 1682993862.988830 (MatMul_93)
#541) after  onnx.MatMul Time elapsed: 0.076849 accumulated: 1682993863.065679 (MatMul_93)
```

### CODE REVIEW

File: src/Conversion/KrnlToLLVM/KrnlInstrument.cpp [hunk 1]
```diff
@@ -59,8 +59,8 @@ class KrnlInstrumentOpLowering : public ConversionPattern {
     Value tag = create.llvm.constant(
         IntegerType::get(context, 64), (int64_t)instrumentOp.getTag());
     StringRef nodeName;
-    if (instrumentOp.getNodeName().has_value())
-      nodeName = instrumentOp.getNodeName().value();
+    if (auto nameLoc = loc.dyn_cast<NameLoc>())
+      nodeName = nameLoc.getName();
```
Comment (tungld) [inline src/Conversion/KrnlToLLVM/KrnlInstrument.cpp:63]: Could we try to get from both NodeName and NameLoc? Is there any case that we have NodeName but not NameLoc and vice versa?

Comment (imaihal) [reply src/Conversion/KrnlToLLVM/KrnlInstrument.cpp:63]: Node name in onnx model is set in both `onnx_node_name`attribute (NodeName) and location (NameLoc) when importing the model. The Nodename is often lost by rewriting, but the NameLoc is not lost. So, there are the case that it has NameLoc, but not NodeName.  I think it is better to use NameLoc here.

Comment (chentong319) [reply src/Conversion/KrnlToLLVM/KrnlInstrument.cpp:63]: The node_name attribute is not the predefined attribute for ONNX Ops, and will be lost by our rewriting.

Comment (imaihal) [reply src/Conversion/KrnlToLLVM/KrnlInstrument.cpp:63]: I see. I will add the cases. Thanks!

Comment (imaihal) [reply src/Conversion/KrnlToLLVM/KrnlInstrument.cpp:65]: I added the case for FusedLoc and FileLineColLoc. In the FusedLoc case, name of each location are combined and the combined name is used as nodeName. In the FileLineColLoc,  filename and fileline are combined and the combined name is used.

File: src/Conversion/KrnlToLLVM/KrnlInstrument.cpp [hunk 2]
```diff
@@ -61,7 +61,28 @@ class KrnlInstrumentOpLowering : public ConversionPattern {
...
       nodeName = instrumentOp.getNodeName().value();
-    else
+    else if (auto nameLoc = loc.dyn_cast<NameLoc>())
+      nodeName = nameLoc.getName();
+    else if (auto fusedLoc = loc.dyn_cast<FusedLoc>()) {
+      // Combine each location name and set it as nodeName.
+      std::string name;
+      for (Location locIt : fusedLoc.getLocations()) {
+        if (auto nameLocIt = locIt.dyn_cast<NameLoc>())
+          name += nameLocIt.getName().str() + "-";
```
Comment (tungld) [inline src/Conversion/KrnlToLLVM/KrnlInstrument.cpp:71]: Could we use `rewriter.getFusedLoc({nameLocIt, locIT})` instead?

Comment (tungld) [reply src/Conversion/KrnlToLLVM/KrnlInstrument.cpp:71]: Oh, it is already a FusedLoc. Please ignore my previous comment.

File: src/Conversion/KrnlToLLVM/KrnlInstrument.cpp [hunk 3]
```diff
@@ -61,7 +61,28 @@ class KrnlInstrumentOpLowering : public ConversionPattern {
...
+        else if (auto fileLineColLoc = locIt.dyn_cast<FileLineColLoc>())
+          name += fileLineColLoc.getFilename().str() + "-";
+      }
+      if (name.empty())
+        name = "NOTSET";
+      name.pop_back(); // remove last "-"
```
Comment (tungld) [inline src/Conversion/KrnlToLLVM/KrnlInstrument.cpp:77]: This will erase "T" in case of "NOTSET". Should be covered by `else`?

Comment (imaihal) [reply src/Conversion/KrnlToLLVM/KrnlInstrument.cpp]: Oops. Thanks for the comment.

**Comment by imaihal** [comment]
@jenkins-droid test this please

**Comment by chentong319** [comment]
Could you try to compile a .mlir file and show the result?

**Comment by imaihal** [comment]
@chentong319 
>Could you try to compile a .mlir file and show the result?

Filename and line are shown as follows when node name is lost.
```
#516) before onnx.Concat Time elapsed: 0.000000 accumulated: 1683005951.302567 (Concat_74)
#517) after  onnx.Concat Time elapsed: 0.000001 accumulated: 1683005951.302568 (Concat_74)
#518) before onnx.Reshape Time elapsed: 0.000001 accumulated: 1683005951.302569 (Reshape_75)
#519) after  onnx.Reshape Time elapsed: 0.000000 accumulated: 1683005951.302569 (Reshape_75)
#520) before onnx.Dim Time elapsed: 0.000001 accumulated: 1683005951.302570 (roberta-base-11.onnx.mlir:362)
#521) after  onnx.Dim Time elapsed: 0.000001 accumulated: 1683005951.302571 (roberta-base-11.onnx.mlir:362)
#522) before onnx.Dim Time elapsed: 0.000000 accumulated: 1683005951.302571 (roberta-base-11.onnx.mlir:365)
#523) after  onnx.Dim Time elapsed: 0.000001 accumulated: 1683005951.302572 (roberta-base-11.onnx.mlir:365)
#524) before onnx.Concat Time elapsed: 0.000001 accumulated: 1683005951.302573 (Concat_84)
#525) after  onnx.Concat Time elapsed: 0.000000 accumulated: 1683005951.302573 (Concat_84)
#526) before onnx.Reshape Time elapsed: 0.000001 accumulated: 1683005951.302574 (Reshape_85)
#527) after  onnx.Reshape Time elapsed: 0.000001 accumulated: 1683005951.302575 (Reshape_85)
#528) before onnx.Transpose Time elapsed: 0.000000 accumulated: 1683005951.302575 (Transpose_86)
#529) after  onnx.Transpose Time elapsed: 0.000040 accumulated: 1683005951.302615 (Transpose_86)
#530) before onnx.Transpose Time elapsed: 0.000001 accumulated: 1683005951.302616 (Transpose_87)
#531) after  onnx.Transpose Time elapsed: 0.000705 accumulated: 1683005951.303321 (Transpose_87)
#532) before onnx.MatMul Time elapsed: 0.000001 accumulated: 1683005951.303322 (MatMul_88)
#533) after  onnx.MatMul Time elapsed: 0.077532 accumulated: 1683005951.380854 (MatMul_88)
#534) before onnx.Div Time elapsed: 0.000001 accumulated: 1683005951.380855 (Div_90)
#535) after  onnx.Div Time elapsed: 0.000608 accumulated: 1683005951.381463 (Div_90)
#536) before onnx.Add Time elapsed: 0.000001 accumulated: 1683005951.381464 (Add_91)
#537) after  onnx.Add Time elapsed: 0.001257 accumulated: 1683005951.382721 (Add_91)
#538) before onnx.Softmax Time elapsed: 0.000000 accumulated: 1683005951.382721 (roberta-base-11.onnx.mlir:378)
#539) after  onnx.Softmax Time elapsed: 0.006424 accumulated: 1683005951.389145 (roberta-base-11.onnx.mlir:378)
#540) before onnx.MatMul Time elapsed: 0.000001 accumulated: 1683005951.389146 (MatMul_93)
#541) after  onnx.MatMul Time elapsed: 0.076896 accumulated: 1683005951.466042 (MatMul_93)
#542) before onnx.Transpose Time elapsed: 0.000001 accumulated: 1683005951.466043 (Transpose_94)
#543) after  onnx.Transpose Time elapsed: 0.000034 accumulated: 1683005951.466077 (Transpose_94)
#544) before onnx.Dim Time elapsed: 0.000000 accumulated: 1683005951.466077 (roberta-base-11.onnx.mlir:381)
#545) after  onnx.Dim Time elapsed: 0.000001 accumulated: 1683005951.466078 (roberta-base-11.onnx.mlir:381)
#546) before onnx.Dim Time elapsed: 0.000001 accumulated: 1683005951.466079 (roberta-base-11.onnx.mlir:384)
#547) after  onnx.Dim Time elapsed: 0.000001 accumulated: 1683005951.466080 (roberta-base-11.onnx.mlir:384)
#548) before onnx.Concat Time elapsed: 0.000000 accumulated: 1683005951.466080 (Concat_103)
#549) after  onnx.Concat Time elapsed: 0.000001 accumulated: 1683005951.466081 (Concat_103)
#550) before onnx.Reshape Time elapsed: 0.000003 accumulated: 1683005951.466084 (Reshape_104)
#551) after  onnx.Reshape Time elapsed: 0.000001 accumulated: 1683005951.466085 (Reshape_104)
#552) before onnx.MatMul Time elapsed: 0.000001 accumulated: 1683005951.466086 (MatMul_105)
#553) after  onnx.MatMul Time elapsed: 0.019767 accumulated: 1683005951.485853 (MatMul_105)
```

> What will happen if you dump the .mlir file with --preserveLocations and then compile the .mlir file?

I can see following result by --EmitZHigIR with --preserveLocations. Locations are added also in krnl.runtime_instrument().  
``` 
   "krnl.runtime_instrument"() {opName = "onnx.MatMul", tag = 5 : i64} : () -> () loc(#loc1)
    %164 = "zhigh.Stick"(%160) {layout = "3DS"} : (tensor<?x?x64xf32>) -> tensor<?x?x64xf32, #zhigh.layout<{dataLayout = "3DS"}>> loc(#loc1)
    %165 = "zhigh.Stick"(%163) {layout = "3DS"} : (tensor<?x64x?xf32>) -> tensor<?x64x?xf32, #zhigh.layout<{dataLayout = "3DS"}>> loc(#loc1)
    %166 = "zhigh.MatMul"(%164, %165, %0) : (tensor<?x?x64xf32, #zhigh.layout<{dataLayout = "3DS"}>>, tensor<?x64x?xf32, #zhigh.layout<{dataLayout = "3DS"}>>, none) -> tensor<?x?x?xf32, #zhigh.layout<{da\
taLayout = "3DS"}>> loc(#loc1)
    %167 = "zhigh.Unstick"(%166) : (tensor<?x?x?xf32, #zhigh.layout<{dataLayout = "3DS"}>>) -> tensor<?x?x?xf32> loc(#loc1)
    "krnl.runtime_instrument"() {opName = "onnx.MatMul", tag = 6 : i64} : () -> () loc(#loc1)
    "krnl.runtime_instrument"() {opName = "onnx.Dim", tag = 5 : i64} : () -> () loc(#loc1)
    %168 = "onnx.Dim"(%147) {axis = 0 : si64} : (tensor<?x12x?x64xf32>) -> tensor<1xi64> loc(#loc1)
    "krnl.runtime_instrument"() {opName = "onnx.Dim", tag = 6 : i64} : () -> () loc(#loc1)
    "krnl.runtime_instrument"() {opName = "onnx.Dim", tag = 5 : i64} : () -> () loc(#loc1)
    "krnl.runtime_instrument"() {opName = "onnx.Dim", tag = 6 : i64} : () -> () loc(#loc1)
    "krnl.runtime_instrument"() {opName = "onnx.Dim", tag = 5 : i64} : () -> () loc(#loc1)
    "krnl.runtime_instrument"() {opName = "onnx.Dim", tag = 6 : i64} : () -> () loc(#loc1)
    "krnl.runtime_instrument"() {opName = "onnx.Concat", tag = 5 : i64} : () -> () loc(#loc1)
    %169 = "onnx.Concat"(%168, %91, %158, %161) {axis = 0 : si64} : (tensor<1xi64>, tensor<1xi64>, tensor<1xi64>, tensor<1xi64>) -> tensor<4xi64> loc(#loc1)
    "krnl.runtime_instrument"() {opName = "onnx.Concat", tag = 6 : i64} : () -> () loc(#loc1)
    "krnl.runtime_instrument"() {opName = "onnx.Reshape", tag = 5 : i64} : () -> () loc(#loc1)
    %170 = "onnx.Reshape"(%167, %169) {allowzero = 0 : si64} : (tensor<?x?x?xf32>, tensor<4xi64>) -> tensor<?x12x?x?xf32> loc(#loc1)
    "krnl.runtime_instrument"() {opName = "onnx.Reshape", tag = 6 : i64} : () -> () loc(#loc1)
    "krnl.runtime_instrument"() {nodeName = "Div_90", opName = "onnx.Div", tag = 5 : i64} : () -> () loc(#loc336)
    %171 = "onnx.Div"(%170, %60) {onnx_node_name = "Div_90"} : (tensor<?x12x?x?xf32>, tensor<f32>) -> tensor<?x12x?x?xf32> loc(#loc336)
    "krnl.runtime_instrument"() {nodeName = "Div_90", opName = "onnx.Div", tag = 6 : i64} : () -> () loc(#loc336)
    "krnl.runtime_instrument"() {nodeName = "Add_91", opName = "onnx.Add", tag = 5 : i64} : () -> () loc(#loc337)
    %172 = "onnx.Add"(%171, %102) {onnx_node_name = "Add_91"} : (tensor<?x12x?x?xf32>, tensor<?x1x1x?xf32>) -> tensor<?x12x?x?xf32> loc(#loc337)
    "krnl.runtime_instrument"() {nodeName = "Add_91", opName = "onnx.Add", tag = 6 : i64} : () -> () loc(#loc337)
    "krnl.runtime_instrument"() {opName = "onnx.Dim", tag = 5 : i64} : () -> () loc(#loc338)
    %173 = "onnx.Dim"(%172) {axis = 2 : si64} : (tensor<?x12x?x?xf32>) -> tensor<1xi64> loc(#loc338)
    "krnl.runtime_instrument"() {opName = "onnx.Dim", tag = 6 : i64} : () -> () loc(#loc338)
    "krnl.runtime_instrument"() {opName = "onnx.Dim", tag = 5 : i64} : () -> () loc(#loc338)
    %174 = "onnx.Dim"(%172) {axis = 3 : si64} : (tensor<?x12x?x?xf32>) -> tensor<1xi64> loc(#loc338)
    "krnl.runtime_instrument"() {opName = "onnx.Dim", tag = 6 : i64} : () -> () loc(#loc338)
    "krnl.runtime_instrument"() {opName = "onnx.Concat", tag = 5 : i64} : () -> () loc(#loc338)
    %175 = "onnx.Concat"(%90, %173, %174) {axis = 0 : si64} : (tensor<1xi64>, tensor<1xi64>, tensor<1xi64>) -> tensor<3xi64> loc(#loc338)
    "krnl.runtime_instrument"() {opName = "onnx.Concat", tag = 6 : i64} : () -> () loc(#loc338)
    "krnl.runtime_instrument"() {opName = "onnx.Reshape", tag = 5 : i64} : () -> () loc(#loc338)
    %176 = "onnx.Reshape"(%172, %175) {allowzero = 0 : si64} : (tensor<?x12x?x?xf32>, tensor<3xi64>) -> tensor<?x?x?xf32> loc(#loc338)
    "krnl.runtime_instrument"() {opName = "onnx.Reshape", tag = 6 : i64} : () -> () loc(#loc338)
    "krnl.runtime_instrument"() {opName = "onnx.Softmax", tag = 5 : i64} : () -> () loc(#loc338)
    %177 = "zhigh.Stick"(%176) {layout = "3DS"} : (tensor<?x?x?xf32>) -> tensor<?x?x?xf32, #zhigh.layout<{dataLayout = "3DS"}>> loc(#loc338)
    %178 = "zhigh.Softmax"(%177) {act_func = "ACT_NONE"} : (tensor<?x?x?xf32, #zhigh.layout<{dataLayout = "3DS"}>>) -> tensor<?x?x?xf32, #zhigh.layout<{dataLayout = "3DS"}>> loc(#loc338)
    %179 = "zhigh.Unstick"(%178) : (tensor<?x?x?xf32, #zhigh.layout<{dataLayout = "3DS"}>>) -> tensor<?x?x?xf32> loc(#loc338)
    "krnl.runtime_instrument"() {opName = "onnx.Softmax", tag = 6 : i64} : () -> () loc(#loc338)


 :
 :
 :
#loc1 = loc("MatMul_88")
#loc336 = loc("Div_90")
#loc337 = loc("Add_91")
#loc338 = loc("Softmax_92")

```

**Comment by AlexandreEichenberger** [comment]
@chentong319 do you have further questions, suggestions that you would like to be implemented here?
@imaihal restarted the build to see if the build failures are now fixed

**Comment by imaihal** [comment]
@AlexandreEichenberger @chentong319 @tungld All tests are passed. Do you have other comments?

File: src/Conversion/KrnlToLLVM/KrnlInstrument.cpp [hunk 4]
```diff
@@ -87,6 +111,21 @@ class KrnlInstrumentOpLowering : public ConversionPattern {
         StringRef("OMInstrumentPoint"), llvmVoidTy,
         {opaquePtrTy, llvmI64Ty, opaquePtrTy});
   }
+
+  // Extract file name from path
+  // Using std::filesystem::path::filename in C++17 is smart, but C++17 not
+  // supported in Jenkins test on some platforms now.
+  std::string getFileName(std::string path) const {
```
Comment (tungld) [inline src/Conversion/KrnlToLLVM/KrnlInstrument.cpp:118]: How about using `llvm::sys::path::filename(inputFilename)`? https://llvm.org/doxygen/namespacellvm_1_1sys_1_1path.html#a5ab6bf136f12ce0bfe2a9754cf1644c7

Comment (imaihal) [reply src/Conversion/KrnlToLLVM/KrnlInstrument.cpp]: Done. Thanks!

**Review by chentong319 — APPROVED** [review]
LGTM!

**Pull request merged**